No more app domains in ATT framework

danielmarbach opened the pull request:
As discussed in Budapest I gave it a quick go to get rid of the AppDomains in the ATT Framework. 

This showed one really nasty thing we are currently doing. In order to move a message into the error queue we corrupt the header serializer (which is a static field) and then let the test run. I tried to give it a go by making the corruptor scoped so that you can rollback those changes. Unfortunately this doesn't work because in case of a forwarding to the error queue in the MsmqReceiveBehaviors the pipeline is aborted. Making the header serialization/deserialization part of the `MsmqUtilities` virtual and use an injected instance is also not an option because then you need to add `InternalsVisibleTo` from the Core to the ATTs. This raised the following design concern (not sure if it is a general issue or just in the MSMQ case):

The ReceiveBehavior should be wrapped with a behavior which catches a specific type of exception and the triggers the forwarding to the error queue. Or should we even have a seperate pipeline for that?

@mauroservienti mentioned that for integration testing purposes having the ability to spin up AppDomains would be still nice to have. We decided to remove it anyway and agreed that probably we need a specialized ATT framework for integration testing. This is out of scope for this PR.

@Particular/nservicebus thoughts?

mauroservienti commented:
Had an interesting discussion yesterday with @danielmarbach regarding this, I agree and I am :+1: in removing the additional complexity of the AppDomain thing, especially if it adds no value.

One thing to keep in mind is that we need to find a way to run end-to-end tests capable to discover things such as #2778 that need to load multiple versions of the same assembly, or scenarios where we want to test communication between endpoints running different versions of NServiceBus and/or of the same transport.

There is a larger scope (tracked by https://github.com/Particular/FeatureDevelopment/issues/340) that is out of the scope of this PR, so I'm :+1: to merge

adamralph commented:
Is there any functional change resulting from this? E.g. fixing intermittent errors, improving performance etc., or is it purely an internal refactoring with no visible change from the outside?

danielmarbach commented:
@mauroservienti this is not ready to be merged

@adamralph currently it is a cleanup/internal refactoring. Which raised the
design question in the PR description. Depending on the outcome of this
discussion it might have visible changes.

Perf improvement is small because only a handful of tests were spinning up
appdomains

>

adamralph commented:
@danielmarbach ok, cool. I was wondering if there were some issues we could close (raise first if required) as part of it but if it's just internal refactoring then that's not the case.

WRT the design change, is that something we should tackle separately? I.e. does this PR make that situation worse, or has just made you aware of the pre-existing problem?

mauroservienti commented:
> this is not ready to be merged 

I know, the intention was a +1 to the idea :-)

danielmarbach commented:
@adamralph all at once. The PR would make it worse. Because with removed
AppDomains the previous hack would corrupt the serializer for all tests
executed after it. Of course I could hack in a try/finally into those ATTs
but the original design question remains. Try/finally would allow this PR
to be merged but there is still a likelyhood that any ATT executed in
parallel will randomly fail

>

adamralph commented:
@danielmarbach I see, makes sense.

Question about the ATT framework. IIRC this is pushed out as https://www.nuget.org/packages/NServiceBus.AcceptanceTesting/ - is that correct?

If so, what is the consequence of breaking changes to public types?

danielmarbach commented:
@adamralph Up until we haven't treated this with SemVer because we never
officially released it, it's not documented... Probably what we should
discuss is why we are posting it to nuget.

SzymonPobiega commented:
Is it just me or are these tests now executed **much** faster?

SzymonPobiega commented:
@danielmarbach I've refactored the problematic `Cant_convert` tests using plain `System.Messaging` API. :heavy_minus_sign: is there is magic string representing a queue name. :heavy_plus_sign: is that we no longer have this funky serializer corruption code. 

Also I've moved it a step further and now the `ScenatioContext` no longer inherits from `ContextBoundObject`. I made it possible to fix the context logging so that now all the log messages are appended to a collection inside the scenario context and can be asserted on.

danielmarbach commented:
Awesome!

Am Samstag, 25. Juli 2015 schrieb Szymon Pobiega :

> @danielmarbach <https://github.com/danielmarbach> I've refactored the
> problematic Cant_convert tests using plain System.Messaging API. [image:
> :heavy_minus_sign:] is there is magic string representing a queue name. [image:
> :heavy_plus_sign:] is that we no longer have this funky serializer
> corruption code.
>
> Also I've moved it a step further and now the ScenatioContext no longer
> inherits from ContextBoundObject. I made it possible to fix the context
> logging so that now all the log messages are appended to a collection
> inside the scenario context and can be asserted on.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2781#issuecomment-124846797>
> .
>

SzymonPobiega commented:
I've rebased it on top of develop. Since the serializer corruption is solved, is this one ready to be merged?

danielmarbach commented:
Let me review and I'll merge if good

Am Montag, 27. Juli 2015 schrieb Szymon Pobiega :

> I've rebased it on top of develop. Since the serializer corruption is
> solved, is this one ready to be merged?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2781#issuecomment-125167371>
> .
>

danielmarbach commented:
@SzymonPobiega pushed a few changes. If OK for you go ahead and merge

SzymonPobiega commented on src/NServiceBus.AcceptanceTesting/ContextAppender.cs:
```diff
@@ -0,0 +1,171 @@
+        {
+            lock (contextLocker)
+            {
+                context = newContext;
```
Why do we need a lock for an assignment?

danielmarbach replied:
I think you are right. It is a leftover from my thinking process about parallel execution. Shall I fix it?

SzymonPobiega replied:
I'll do it. Just wanted to make sure.

SimonCropp commented on src/NServiceBus.AcceptanceTests/Sagas/When_receiving_that_completes_the_saga.cs:
```diff
@@ -58,6 +58,7 @@ public void Should_ignore_messages_afterwards()
                         SomeId = c.Id
                     }));
                 })
+                .AllowExceptions()
```
why?

SzymonPobiega replied:
Good catch. I blindly added it because the test was throwing but I *think* the problem was an unnecessary `IHandleSagaNotFound`
